### PR TITLE
Avoid cartesian products in pivot csv exports

### LINKED ISSUE #49977: Exporting pivots to XLSX with "Keep Data Pivoted" selected can Crash Metabase (Heap memory error) and pegs CPU
### Describe the bug

Exporting Pivots to XLSX can eat memory and can cause the JVM to crash. 

### To Reproduce

1. Set up v51.1 locally with 1GB RAM
2. Create a pivot with accounts and invoices that has a handful of breakouts (Loom recording below)
3. Export to XLSX with "Keep Data Pivoted" selected
4. Metabase crashed with OOM error

https://www.loom.com/share/9446d6efe8af4bba8ebc23b322c8d3e6

### Expected behavior

We [recommend deploying with 1G](https://www.metabase.com/learn/metabase-basics/administration/administration-and-operation/metabase-in-production#metabase-application-server-size) for every 20 concurrent users so I wouldn't expect to to hit heap memory in this case. The pivot table wasn't very large:

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres",
      "sqlserver",
      "mysql",
      "oracle",
      "databricks"
    ],
    "run-mode": "prod",
    "plan-alias": "enterprise-self-hosted",
    "version": {
      "date": "2024-10-23",
      "tag": "v1.51.1",
      "hash": "b8178a8"
    },
    "settings": {
      "report-timezone": null
    },
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "15.1 (Debian 15.1-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    }
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.24+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.24",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.24+8",
    "os.name": "Linux",
    "os.version": "5.15.153.1-microsoft-standard-WSL2",
    "user.language": "en",
    "user.timezone": "GMT"
  }
}
```

### Severity

blocking upgrade

### Additional context

Customer that initially reported this is running on 2GB and sees their instances crash intermittently with pivot exports. They want to upgrade to the next minor release to see some other patches but they're concerned about moving to a version where exporting these files pivoted is the default behavior given these crashes.

### PR DESCRIPTION
**Pull request by noahmoss**
Likely resolves https://github.com/metabase/metabase/issues/49977

This PR optimizes pivoted CSV exports by avoiding generating Cartesian products of all row and column values and instead building the result rows and columns as tree structures from the actual data. These are accumulated in the `:row-paths` and `:col-paths` keys in the pivot data structure returned by `init-pivot`.

This mirrors the approach taken by the frontend, which doesn't see equivalent performance issues as the backend: https://github.com/metabase/metabase/blob/abad978404a78c111f14414648c401636f0b7dad/frontend/src/metabase/lib/data_grid.js#L53-L55

The trees are built as nested `ordered-map`s (with `ordered-set`s to accumulate the leaf values) so that we don't lose the ordering from the original data, since this is the order of the output rows if no alternate ordering is provided. (I'm not sure if this is the best default behavior, since this ordering is pretty arbitrary, but for now I'm just trying to match the frontend).

Then, in `sort-path-tree` we traverse the tree and convert each level to a `sorted-map` or `sorted-set` as needed, based on the pivot settings. Finally, we enumerate every path from the root to a leaf to get the ultimate `row-combos` and `col-combos` which manifest as actual rows/columns in the CSV.

I've also made some tweaks to the surrounding code to ensure that column and row ordering is respected, and to handle some other edge cases. And I've added some unit tests for new helper functions, and tweaked the expected output of other tests to be more correct. There's a lot more cleanup and further optimization that can be done here but I'm trying to keep the changes relatively focused so that we can resolve the main source of perf problems.

### CODE REVIEW

**Comment by noahmoss** [comment]
* **#52551** <a href="https://app.graphite.dev/github/pr/metabase/metabase/52551?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/52551?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/metabase/query_processor/pivot/postprocess.clj [hunk 1]
```diff
@@ -99,16 +100,28 @@
...
-  (update m k conj v))
+(defn- assoc-in-path-tree
+  "Assocs a value in a path tree, which should consist of a hierarchy of ordered-maps, with leaf values stored in
+  ordered-sets."
+  [tree ks v]
+  (let [step (fn step [tree [k & ks]]
+               (if ks
+                 (if-let [next-map (get tree k)]
+                   (assoc tree k (step next-map ks))
+                   (assoc tree k (step (ordered-map/ordered-map) ks)))
```
Comment (alexander-yakushev) [inline src/metabase/query_processor/pivot/postprocess.clj:120]: Can probably simplify this to `(assoc tree k (step (or (get tree k) (ordered-map/ordered-map)) ks))`? Or if it's too mouthful, then
```clj
(let [next-map (or (get tree k) (ordered-map/ordered-map))]
  (assoc tree k (step next-map ks)))
```

File: src/metabase/query_processor/pivot/postprocess.clj [hunk 2]
```diff
@@ -99,16 +100,28 @@
...
+                 (if k
+                   (assoc tree k (conj (get tree k (ordered-set/ordered-set)) v))
```
Comment (alexander-yakushev) [inline src/metabase/query_processor/pivot/postprocess.clj:122]: You don't want to use the 3-arg `get` arity here because the default value is computed every time. `or` is better here.

File: src/metabase/query_processor/pivot/postprocess.clj [hunk 3]
```diff
@@ -195,6 +208,10 @@
         (update :row-count (fn [v] (if v (inc v) 0)))
         (update :data update-in (concat row-path col-path)
                 #(update-aggregate (or % (zipmap pivot-measures (repeat {}))) measure-vals measures))
+        (update :row-paths
+                #(when (seq row-path) (assoc-in-path-tree % (butlast row-path) (last row-path))))
+        (update :col-paths
+                #(when (seq col-path) (assoc-in-path-tree % (butlast col-path) (last col-path))))
```
Comment (alexander-yakushev) [inline src/metabase/query_processor/pivot/postprocess.clj:214]: `butlast` is usually a bad sign, but I can look at it during the next iteration, doesn't need to be addressed outright.

**Comment by alexander-yakushev** [comment]
Amazing job diffusing the cartesian explosion! I left a couple of comments. It looks like this will significantly reduce the memory watermark during the export. Once this PR is merged, I'll do the measurements and add my improvements on top.

File: src/metabase/query_processor/pivot/postprocess.clj (hunk 3 above)
Comment (noahmoss) [reply src/metabase/query_processor/pivot/postprocess.clj]: Refactored the code to avoid it

**Review by alexander-yakushev — APPROVED** [review]
